chains/evm/voter/voter.go

```diff
@@ -28,42 +44,147 @@ type MessageHandler interface {
 }
 
 type EVMVoter struct {
-	mh             MessageHandler
-	client         ChainClient
-	fabric         calls.TxFabric
-	gasPriceClient calls.GasPricer
+	mh                   MessageHandler
+	client               ChainClient
+	fabric               calls.TxFabric
+	gasPriceClient       calls.GasPricer
+	pendingProposalVotes map[common.Hash]uint8
 }
 
-func NewVoter(mh MessageHandler, client ChainClient, fabric calls.TxFabric, gasPriceClient calls.GasPricer) *EVMVoter {
-	return &EVMVoter{
-		mh:             mh,
-		client:         client,
-		fabric:         fabric,
-		gasPriceClient: gasPriceClient,
+func NewVoter(mh MessageHandler, client ChainClient, fabric calls.TxFabric, gasPriceClient calls.GasPricer) (*EVMVoter, error) {
+	voter := &EVMVoter{
+		mh:                   mh,
+		client:               client,
+		fabric:               fabric,
+		gasPriceClient:       gasPriceClient,
+		pendingProposalVotes: make(map[common.Hash]uint8),
+	}
+
+	ch := make(chan common.Hash)
+	_, err := client.SubscribePendingTransactions(context.TODO(), ch)
+	if err != nil {
+		return nil, err
 	}
+
+	go voter.trackProposalPendingVotes(ch)
+	return voter, nil
 }
 
+// VoteProposal checks if relayer already voted and is threshold
+// satisfied and casts a vote if it isn't
 func (v *EVMVoter) VoteProposal(m *relayer.Message) error {
 	prop, err := v.mh.HandleMessage(m)
 	if err != nil {
 		return err
 	}
+
+	votedByTheRelayer, err := calls.IsProposalVotedBy(v.client, v.client.RelayerAddress(), prop)
+	if err != nil {
+		return err
+	}
+	if votedByTheRelayer {
+		return nil
+	}
+
+	shouldVoteChn := make(chan bool)
+	go v.shouldVoteForProposal(shouldVoteChn, prop, 0)
+
+	shouldVote := <-shouldVoteChn
```

Comment:
Since this line is blocking i do not see any reasons of using goroutines on line 90 as well as channels. It can be done just with simple blocking function call. 
I think that in this case its over-complication that decreasing code maintainability, readability and may lead to unexpected bugs + memory leaks in future.

Reply:
Yep, initial implementation was a bit different so I forgot to remove it.